Adding route dependencies module.

rhysrevans3 opened the pull request:
**Description:**

Generate a list of route dependencies from the environment variable `STAC_FASTAPI_ROUTE_DEPENDENCIES` or the file that this environment variable points to. Feed these route dependencies into the `STACApi` on initialisation. 

**PR Checklist:**

- [x] Code is formatted and linted (run `pre-commit run --all-files`)
- [x] Tests pass (run `make test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] Changes are added to the changelog

rhysrevans3 commented:
@jonhealy1 sorry I've been at a conference so have only just caught up on the authentication work. The basic auth looks great. Am I right in saying that all endpoints are now closed unless they're included in the public_endpoints list in the `BASIC_AUTH` environment variable?

jonhealy1 commented:
@pedro-cf Hi Pedro. Are you able to look at this pr and respond to @rhysrevans3? I think you would do a better job with this. Thanks!

pedro-cf commented:
> @jonhealy1 sorry I've been at a conference so have only just caught up on the authentication work. The basic auth looks great. Am I right in saying that all endpoints are now closed unless they're included in the public_endpoints list in the `BASIC_AUTH` environment variable?

Greetings, only if `BASIC_AUTH` is enabled will the endpoints be protected, unless under the `public_endpoints` list.

pedro-cf commented:
> @jonhealy1 sorry I've been at a conference so have only just caught up on the authentication work. The basic auth looks great. Am I right in saying that all endpoints are now closed unless they're included in the public_endpoints list in the `BASIC_AUTH` environment variable?

Greetings, 

if `BASIC_AUTH` is enabled the endpoints outside of `public_endpoints` will be protected.
`BASIC_AUTH` is optional and off by default. You enable it by setting the `BASIC_AUTH` env variable.

rhysrevans3 commented:
@jonhealy1 @pedro-cf thanks both I think I understand now. Could I get your opinion on this pull request as it also deals with auth. Are you happy that this and basic auth can co-exist? Or would you like me to try and merge the two? I currently don't have a way to protect all endpoints without explicitly writing them all down. You could add wildcards but this would probably need changes to https://github.com/stac-utils/stac-fastapi

jonhealy1 commented:
I definitely like the idea of supporting OAuth2.

jonhealy1 commented:
Wondering about the best way to integrate this with basic auth? This will need instructions added to the readme as well and a docker-compose demo file I think.

rhysrevans3 commented:
I've been thinking the same thing I will add instructions to the readme and a docker-compose demo file. They do currently work together but it might be confusing to have both in their current implementation.

I could update basic auth to use this pull requests methods but it would change the config and would remove the `*` option. I could add the star option to this pull request but that would require changes to [stac-fastapi](https://github.com/stac-utils/stac-fastapi).

another thought I had was this would fit better in the [stac-fastapi](https://github.com/stac-utils/stac-fastapi) so it can be used in all of the backends?

rhysrevans3 commented:
I've made a [pull request](https://github.com/stac-utils/stac-fastapi-elasticsearch-opensearch/pull/251) in stac-fastapi that would allow for default route dependencies to be added. This would make the merging of basic auth and this pull request much simpler.

rhysrevans3 commented:
@jonhealy1 @pedro-cf I've updated basic auth to use the route dependencies method. The configuration syntax for basic auth has changed.

Basic auth can now be used as a dependency within route dependencies.

The example in the readme now looks like:
```
[
  {
    "routes": [
      {
        "method": "*",
        "path": "*"
      }
    ],
    "dependencies": [
      {
        "method": "stac_fastapi.core.basic_auth.BasicAuth",
        "kwargs": {
          "credentials":[
            {
              "username": "admin",
              "password": "admin"
            }
          ]
        }
      }
    ]
  },
  {
    "routes": [
      {"path": "/", "method": ["GET"]},
      {"path": "/conformance", "method": ["GET"]},
      {"path": "/collections/{collection_id}/items/{item_id}", "method": ["GET"]},
      {"path": "/search", "method": ["GET", "POST"]},
      {"path": "/collections", "method": ["GET"]},
      {"path": "/collections/{collection_id}", "method": ["GET"]},
      {"path": "/collections/{collection_id}/items", "method": ["GET"]},
      {"path": "/queryables", "method": ["GET"]},
      {"path": "/queryables/collections/{collection_id}/queryables", "method": ["GET"]},
      {"path": "/_mgmt/ping", "method": ["GET"]}
    ],
    "dependencies": [
      {
        "method": "stac_fastapi.core.basic_auth.BasicAuth",
        "kwargs": {
          "credentials":[
            {
              "username": "reader",
              "password": "reader"
            }
          ]
        }
      }
    ]
  }
]
```
I've also removed `public_endpoints`. The default is for all endpoints to be public and `*` can be used for a route's `path` and/or `method` to add a default dependency.

pedro-cf reviewed (commented):
Good work, I would suggest:

- add jsonschema validation for the `route_dependencies.json`
- add real complete working examples `route_dependencies.json`

rhysrevans3 commented:
> Good work, I would suggest:
> 
> * add jsonschema validation for the `route_dependencies.json`
> * add real complete working examples `route_dependencies.json`

I've add some  jsonschema validation for route_dependencies. And added a second docker compose file that use `example/route_dependencies/route_dependencies.json` and gives an example of an Oauth2 route dependency.

jonhealy1 commented on docker-compose.basic_auth_protected.yml:
```diff
@@ -22,7 +22,7 @@ services:
       - ES_VERIFY_CERTS=false
       - BACKEND=elasticsearch
-      - BASIC_AUTH={"users":[{"username":"admin","password":"admin","permissions":"*"},{"username":"reader","password":"reader","permissions":[{"path":"/","method":["GET"]},{"path":"/conformance","method":["GET"]},{"path":"/collections/{collection_id}/items/{item_id}","method":["GET"]},{"path":"/search","method":["GET","POST"]},{"path":"/collections","method":["GET"]},{"path":"/collections/{colle ...
+      - STAC_FASTAPI_ROUTE_DEPENDENCIES=[{"routes":[{"method":"*","path":"*"}],"dependencies":[{"method":"stac_fastapi.core.basic_auth.BasicAuth","kwargs":{"credentials":[{"username":"admin","password":"admin"}]}}]},{"routes":[{"path":"/","method":["GET"]},{"path":"/conformance","method":["GET"]},{"path":"/collections/{collection_id}/items/{item_id}","method":["GET"]},{"path":"/search","method":[ ...
```
Hi. Should the opensearch docker-compose examples change too to STAC_FASTAPI_ROUTE_DEPENDENCIES?

rhysrevans3 replied:
You're right I've updated all the opnesearch ones too now

jonhealy1 commented on README.md:
```diff
@@ -279,75 +279,106 @@ The modified Items with lowercase identifiers will now be visible to users acces
-            "permissions": "*"
+    "dependencies": [
+      {
+        "method": "fastapi.security.OAuth2PasswordBearer",
```
Can you explain this just because I am not savvy. We can or do now support oauth2?

jonhealy1 replied:
If we do - clearly we do - should we add something additional to the changelog about this capability?

rhysrevans3 replied:
The changes I've made allows for any `method` (a class or function) to be used as a route dependency. FastAPI has some of the Oauth2 flows built in which is being used there. You can also implement your own route dependencies. Basic auth has been converted into one so it now doubles up as a example of that too.

jonhealy1 commented on docker-compose.route_dependencies_env.yml:
```diff
@@ -55,7 +55,7 @@ services:
       - ES_VERIFY_CERTS=false
       - BACKEND=opensearch
-      - BASIC_AUTH={"public_endpoints":[{"path":"/","method":"GET"},{"path":"/conformance","method":"GET"},{"path":"/collections/{collection_id}/items/{item_id}","method":"GET"},{"path":"/search","method":"GET"},{"path":"/search","method":"POST"},{"path":"/collections","method":"GET"},{"path":"/collections/{collection_id}","method":"GET"},{"path":"/collections/{collection_id}/items","method":"GET ...
+      - BASIC_AUTH={"users":[{"username":"admin","password":"admin","permissions":"*"},{"username":"reader","password":"reader","permissions":[{"path":"/","method":["GET"]},{"path":"/conformance","method":["GET"]},{"path":"/collections/{collection_id}/items/{item_id}","method":["GET"]},{"path":"/search","method":["GET","POST"]},{"path":"/collections","method":["GET"]},{"path":"/collections/{colle ...
```
Here maybe too

jonhealy1 commented on docker-compose.route_dependencies_file.yml:
```diff
@@ -0,0 +1,129 @@
+      - ES_USE_SSL=false
+      - ES_VERIFY_CERTS=false
+      - BACKEND=opensearch
+      - BASIC_AUTH={"users":[{"username":"admin","password":"admin","permissions":"*"},{"username":"reader","password":"reader","permissions":[{"path":"/","method":["GET"]},{"path":"/conformance","method":["GET"]},{"path":"/collections/{collection_id}/items/{item_id}","method":["GET"]},{"path":"/search","method":["GET","POST"]},{"path":"/collections","method":["GET"]},{"path":"/collections/{colle ...
```
Here as well maybe

jonhealy1 commented on docker-compose.route_dependencies_file.yml:
```diff
@@ -0,0 +1,129 @@
+    volumes:
+      - postgres:/var/lib/postgresql/data
+
+  keycloak:
```
If you could explain keycloak too a little. This is something we can use instead of basic auth?

jonhealy1 replied:
It uses oauth2?

jonhealy1 replied:
Looks super cool btw

rhysrevans3 replied:
[Keycloak](https://www.keycloak.org/) is an identity provider that supports [OIDC](https://www.microsoft.com/en-us/security/business/security-101/what-is-openid-connect-oidc#:~:text=OIDC%20stands%20for%20OpenID%20Connect,information%20without%20compromising%20user%20credentials.) which is an extension to OAuth2.

Basically it handles the user authentication parts. The user authenticates with keycloak then the authorization token they receive can be used by STAC FastAPI to verify they are who they are and their permissions (also via keycloak).

There are some nice diarams explaining the workflow on this [article](https://darutk.medium.com/diagrams-of-all-the-openid-connect-flows-6968e3990660) where STAC FastAPI would be the client and keycloak the server.

Do you think any of this information needs to be added to the readme?

jonhealy1 replied:
Great article. I don't think we need to add any of this information. We could have a separate doc that explains some of this stuff but it would be a bit of work. We could also link the article somewhere maybe?

jonhealy1 commented on examples/route_dependencies/stac-realm.json:
```diff
@@ -0,0 +1,1847 @@
+{
```
If you could explain this too. This is clearly used with keycloak. It's a fairly long file ... but it's probably not a problem - it's only 67 kb.

rhysrevans3 replied:
It's a saved copy of a test keycloak with a realm and one user already created. I could remove this and just add the steps to create a realm and user to the readme if that's preferable?

jonhealy1 replied:
It's nice to have to fire off the docker compose example quickly. It's not being packaged in our pypi release so I don't think the size matters. We could add the steps to create a realm in the readme which would probably be really good.

jonhealy1 replied:
I created an issue on adding documentation for creating a keycloak file

jonhealy1 reviewed (changes requested):
Incredible work here

jonhealy1 commented on CHANGELOG.md:
```diff
@@ -14,6 +14,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ### Changed
```
Maybe something in the Added section about adding examples with keycloak and OAuth2?

rhysrevans3 replied:
done 👍

jonhealy1 commented:
Nice changelog updates thanks!

jonhealy1 commented:
@rhysrevans3 I moved the changelog entries to the Unreleased section in the changelog.

jonhealy1 commented:
Once we release a new version here and on PyPI then we move the changelog entries to the appropriate section which will hopefully be v3.0.0. Waiting for stac-fastapi v3.0.0

jonhealy1 reviewed (approved):
Great work!

pedro-cf commented:
@rhysrevans3 didn't notice but the README needs some correcting:

> Docker Compose Configurations
See docker-compose.basic_auth_protected.yml and docker-compose.basic_auth_public.yml for basic authentication configurations.

jonhealy1 commented:
@pedro-cf Can you open up a new pr to fix this? Good catch.

pedro-cf commented:
> @pedro-cf Can you open up a new pr to fix this? Good catch.

- Would it make sense to simply remove the section? or Add a description for every file? @rhysrevans3  could perhaps help with this?
- Also would it make sense to move the docker-compose files to the examples directory?

jonhealy1 commented:
It does make some sense to move them to the examples folder I think.
